### DISCUSSION TITLE
GRPC Plugin to interact with Porter

### DISCUSSION THREAD

**Discussion by MChorfa** [getporter/porter#1313] [Ideas]
## Backgound

As a non-technical user, I want to perform porter operations and have feedback via a web/desktop application interface

## Related issues

- Rethink interactive commands https://github.com/getporter/porter/issues/1080
- Provide parameters interactively https://github.com/getporter/porter/issues/793
- Unable to edit parameters on Windows https://github.com/getporter/porter/issues/1241
- Make it easier to reference/drop credentials into the bundle directory https://github.com/getporter/porter/issues/496

## Suggestions

- Create a GRPC plugin that exposes the porter commands
- Create a web/desktop client to interact with the GRPC plugin
- Introduce a way to generate parameters and credentials
- Introduce a way to perform OIDC/oauth2 before retrieving values from a third-party provider like (Azure KeyVault, Hashicorp Vault)
- Introduce a way for Porter to provide contextual feedback and progress when executing
- Introduce a way for Porter to provide installations state locally and remote

#### Install

```go

service IntallService {
  rpc Install (IntallRequest) returns (IntallResponse);
}

message IntallRequest {
  string installing = 1;
}

message IntallResponse {
  string output = 1;
}

```

#### Upgrade and streaming output

```go

service UpgradeService {
  rpc Upgrade (stream UpgradeRequest) returns (stream UpgradeResponse);
}

message UpgradeRequest {
  string upgrading = 1;
}

message UpgradeResponse {
  string output = 1;
}

```

### Proposition for interacting and generating (prams and creds)

![image](https://user-images.githubusercontent.com/183869/95004913-99abd600-05bf-11eb-85de-2f3cbbb0a603.png)

### References

- https://github.com/kubenav/kubenav
- https://grpc.io/docs/what-is-grpc/core-concepts/

**Comment by carolynvs** [comment]
🤯 Whoa! I need to digest this and how it would need to be implemented. Very cool idea!

**Comment by carolynvs** [comment]
You brought this up in Slack today and I'd love to chat more about it. I see a lot of potential in having this and had a few thoughts about how someone could go about it:

Make an alternate entrypoint into Porter

Right now we have cmd/porter (cli) and pkg/porter (library with the Porter struct to automate commands). We could add another package, cmd/porterd and make that long running service that responds to gRPC requests. I think that would fit better into Porter's architecture than a plugin.

As you pointed out, it could be used by all sorts of integrations. We could also take advantage of that in the porter operator. The operator would still spawn jobs to execute the invocation image, but could avoid making another job to run the cli and instead send a command to the grpc service.

I'd be happy to help review and collaborate on designs, and a POC for this if you are interested in moving forward with it.

**Reply by MChorfa** [reply]
Yes I would like very much to work on this feature. I will try to draft a PEP very soon

**Comment by bdegeeter** [comment]
This design is inline with goals I've had to POC a service that allows a web/cli interface for a user to see a catalogue of bundles and install.  An SRE would manage some aspect of bundle requirements (like creds) with a subset of parameters exposed to users.  I have a k8s service that does this by watching k8s job resources created by the operator.  I'm currently working on how to extract bundle outputs and provide a subset of them back to the user.  

Down the road, I could see an additional BundleDefintion CRD supporting a collection of known/approved bundles with attributes needed for presentation to a user.  I can help with POC work as well.

Porter supports our separation of responsibility with a design like this.  Anyone who want's an item deployed by the service can create a bundle.  An SRE running the service can manage where items are deployed and what's exposed back to the users.

**Reply by carolynvs** [reply]
Yes, I am very interested in having a set of approved bundles. I started going off on a tangent about having a Bundle Index in the dependencies spec. https://github.com/getporter/proposals/blob/9ed393e424b99f3d2bb09ee1773719dc46278b4f/pep/003-dependency-namespaces-and-labels.md#porter-bundle-index

@bdegeeter if you are interested in helping write or contribute to proposals (not just for this), I would love to have help thinking these through and moving some of them forward.

**Comment by MChorfa** [comment]
@schristoff
